### ISSUE TITLE
IndexMerge can't work as expected when condition has func

### ISSUE THREAD

**Issue by lilinghai** [pingcap/tidb#16409]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
```
1. create table t(a int ,b char(100),c double,key(a),key(a),key(b),key(c));
2. explain select /*+ use_index_merge(t) */ * from t where a < 10 or c < rand();
3. explain select /*+ use_index_merge(t) */ * from t where a < 10 or b < RANDOM_BYTES(10);
```

### 2. What did you expect to see?
IndexMerge is consistent in the 2th and 3th result.
### 3. What did you see instead?
the 2th statement result  
```
+-------------------------+----------+-----------+---------------+----------------------------------------+
| id                      | estRows  | task      | access object | operator info                          |
+-------------------------+----------+-----------+---------------+----------------------------------------+
| Selection_5             | 8000.00  | root      |               | or(lt(d1.t.a, 10), lt(d1.t.c, rand())) |
| └─TableReader_7         | 10000.00 | root      |               | data:TableFullScan_6                   |
|   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:t       | keep order:false, stats:pseudo         |
+-------------------------+----------+-----------+---------------+----------------------------------------+
3 rows in set, 1 warning (0.00 sec)
```
the 3th statement result
```
+-------------------------------+---------+-----------+---------------------+-----------------------------------------------------------+
| id                            | estRows | task      | access object       | operator info                                             |
+-------------------------------+---------+-----------+---------------------+-----------------------------------------------------------+
| IndexMerge_8                  | 6646.67 | root      |                     |                                                           |
| ├─IndexRangeScan_5(Build)     | 3323.33 | cop[tikv] | table:t, index:a(a) | range:[-inf,10), keep order:false, stats:pseudo           |
| ├─IndexRangeScan_6(Build)     | 3323.33 | cop[tikv] | table:t, index:b(b) | range:[-inf,"� $
                                                                                              cU��."), keep order:false, stats:pseudo    |
| └─TableRowIDScan_7(Probe)     | 6646.67 | cop[tikv] | table:t             | keep order:false, stats:pseudo                            |
+-------------------------------+---------+-----------+---------------------+-----------------------------------------------------------+
```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
Release Version: v4.0.0-rc-33-gd69868961
Git Commit Hash: d69868961361126d4788b68fd6cd597adb3ef66c
Git Branch: release-4.0
UTC Build Time: 2020-04-15 02:59:00
GoVersion: go1.13.8
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

**Comment by zz-jason** [comment]
@lzmhhh123 PTAL

**Comment by lzmhhh123** [comment]
I'm afraid this is not a bug. The `rand()` is a deferred expression. It should calculate for each row, rather than build once the value for index scan.

**Comment by eurekaka** [comment]
Yes, according to MySQL's behavior, `rand()` and `random_bytes()` are inconsistent regarding constant folding, `rand()` is evaluated for each row, while `random_bytes()` is evaluated only once at planning stage, so this explain output is as expected.
